--- a/scripts/Makefile
+++ b/scripts/Makefile
@@ -33,8 +33,10 @@ hostprogs-y += unifdef docproc check-lc_

 # These targets are used internally to avoid "is up to date" messages
 PHONY += build_unifdef build_docproc build_check-lc_ctype
-build_unifdef: $(obj)/unifdef
-	@:
+build_unifdef:
+	echo "unifdef done"
+#build_unifdef: $(obj)/unifdef
+#	@:
 build_docproc: $(obj)/docproc
	@:
 build_check-lc_ctype: $(obj)/check-lc_ctype
