--- a/drivers/acpi/Kconfig
+++ b/drivers/acpi/Kconfig
@@ -42,6 +42,9 @@ menuconfig ACPI
 
 if ACPI
 
+config ACPI_MUTEX
+   bool
+
 config ACPI_LEGACY_TABLES_LOOKUP
 	bool
 
--- a/drivers/acpi/acpica/utxfmutex.c
+++ b/drivers/acpi/acpica/utxfmutex.c
@@ -40,6 +40,9 @@
  * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
  * POSSIBILITY OF SUCH DAMAGES.
  */
+#ifdef CONFIG_ACPI_MUTEX
+#define EXPORT_ACPI_INTERFACES
+#endif
 
 #include <acpi/acpi.h>
 #include "accommon.h"
@@ -151,6 +154,10 @@ acpi_acquire_mutex(acpi_handle handle, a
 	return (status);
 }
 
+#ifdef CONFIG_ACPI_MUTEX
+ACPI_EXPORT_SYMBOL(acpi_acquire_mutex)
+#endif
+
 /*******************************************************************************
  *
  * FUNCTION:    acpi_release_mutex
@@ -185,3 +192,7 @@ acpi_status acpi_release_mutex(acpi_hand
 	acpi_os_release_mutex(mutex_obj->mutex.os_mutex);
 	return (AE_OK);
 }
+
+#ifdef CONFIG_ACPI_MUTEX
+ACPI_EXPORT_SYMBOL(acpi_release_mutex)
+#endif
