--- a/drivers/acpi/pci_root.c
+++ b/drivers/acpi/pci_root.c
@@ -494,8 +494,8 @@ static void negotiate_os_control(struct
 			 * it's unsupported. Leave existing configuration
 			 * intact and prevent the OS from touching it.
 			 */
-			dev_info(&device->dev, "FADT indicates ASPM is unsupported, using BIOS configuration\n");
-			*no_aspm = 1;
+			dev_info(&device->dev, "Ignoring BIOS ASPM configuration\n");
+			*no_aspm = 0;
 		}
 	} else {
 		decode_osc_control(root, "OS requested", requested);
--- a/drivers/pci/pci-acpi.c
+++ b/drivers/pci/pci-acpi.c
@@ -863,8 +863,8 @@ static int __init acpi_pci_init(void)
 	}
 
 	if (acpi_gbl_FADT.boot_flags & ACPI_FADT_NO_ASPM) {
-		pr_info("ACPI FADT declares the system doesn't support PCIe ASPM, so disable it\n");
-		pcie_no_aspm();
+		pr_info("ACPI FADT declares the system doesn't support PCIe ASPM, ignore it\n");
+		/* pcie_no_aspm(); */
 	}
 
 	ret = register_acpi_bus_type(&acpi_pci_bus);
