--- a/Makefile
+++ b/Makefile
@@ -263,10 +263,10 @@ CONFIG_SHELL := $(shell if [ -x "$$BASH"
	  else if [ -x /bin/bash ]; then echo /bin/bash; \
	  else echo sh; fi ; fi)

-HOSTCC       = gcc
-HOSTCXX      = g++
-HOSTCFLAGS   = -Wall -Wmissing-prototypes -Wstrict-prototypes -O2 -fomit-frame-pointer -std=gnu89
-HOSTCXXFLAGS = -O2
+HOSTCC       ?= gcc
+HOSTCXX      ?= g++
+HOSTCFLAGS   ?= -Wall -Wmissing-prototypes -Wstrict-prototypes -O2 -fomit-frame-pointer -std=gnu89
+HOSTCXXFLAGS ?= -O2

 ifeq ($(shell $(HOSTCC) -v 2>&1 | grep -c "clang version"), 1)
 HOSTCFLAGS  += -Wno-unused-value -Wno-unused-parameter \
--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -2710,19 +2710,6 @@ config PMC_ATOM
	def_bool y
         depends on PCI

-config VMD
-	depends on PCI_MSI
-	tristate "Volume Management Device Driver"
-	default N
-	---help---
-	  Adds support for the Intel Volume Management Device (VMD). VMD is a
-	  secondary PCI host bridge that allows PCI Express root ports,
-	  and devices attached to them, to be removed from the default
-	  PCI domain and placed within the VMD domain. This provides
-	  more bus resources than are otherwise possible with a
-	  single domain. If you know your system provides one of these and
-	  has devices attached to it, say Y; if you are not sure, say N.
-
 source "net/Kconfig"

 source "drivers/Kconfig"
--- a/drivers/Kconfig
+++ b/drivers/Kconfig
@@ -200,6 +200,19 @@ source "drivers/hwtracing/stm/Kconfig"

 source "drivers/hwtracing/intel_th/Kconfig"

+config VMD
+	depends on PCI_MSI
+	tristate "Volume Management Device Driver"
+	default N
+	---help---
+	  Adds support for the Intel Volume Management Device (VMD). VMD is a
+	  secondary PCI host bridge that allows PCI Express root ports,
+	  and devices attached to them, to be removed from the default
+	  PCI domain and placed within the VMD domain. This provides
+	  more bus resources than are otherwise possible with a
+	  single domain. If you know your system provides one of these and
+	  has devices attached to it, say Y; if you are not sure, say N.
+
 source "drivers/fpga/Kconfig"

 endmenu
--- a/scripts/kconfig/nconf.c
+++ b/scripts/kconfig/nconf.c
@@ -5,7 +5,9 @@
  * Derived from menuconfig.
  *
  */
+#ifndef _GNU_SOURCE
 #define _GNU_SOURCE
+#endif
 #include <string.h>
 #include <stdlib.h>

--- a/scripts/package/builddeb
+++ b/scripts/package/builddeb
@@ -12,6 +12,8 @@

 set -e

+KDEB_SOURCENAME=linux
+
 create_package() {
	local pname="$1" pdir="$2"

@@ -236,24 +238,24 @@ fi
 maintainer="$name <$email>"

 # Try to determine distribution
-if [ -n "$KDEB_CHANGELOG_DIST" ]; then
-        distribution=$KDEB_CHANGELOG_DIST
-elif distribution=$(lsb_release -cs 2>/dev/null) && [ -n "$distribution" ]; then
-        : # nothing to do in this case
-else
-        distribution="unstable"
-        echo >&2 "Using default distribution of 'unstable' in the changelog"
-        echo >&2 "Install lsb-release or set \$KDEB_CHANGELOG_DIST explicitly"
-fi
-
-# Generate a simple changelog template
-cat <<EOF > debian/changelog
-$sourcename ($packageversion) $distribution; urgency=low
-
-  * Custom built Linux kernel.
-
- -- $maintainer  $(date -R)
-EOF
+#if [ -n "$KDEB_CHANGELOG_DIST" ]; then
+#        distribution=$KDEB_CHANGELOG_DIST
+#elif distribution=$(lsb_release -cs 2>/dev/null) && [ -n "$distribution" ]; then
+#        : # nothing to do in this case
+#else
+#        distribution="unstable"
+#        echo >&2 "Using default distribution of 'unstable' in the changelog"
+#        echo >&2 "Install lsb-release or set \$KDEB_CHANGELOG_DIST explicitly"
+#fi
+
+## Generate a simple changelog template
+#cat <<EOF > debian/changelog
+#$sourcename ($packageversion) $distribution; urgency=low
+#
+#  * Custom built Linux kernel.
+#
+# -- $maintainer  $(date -R)
+#EOF

 # Generate copyright file
 cat <<EOF > debian/copyright
@@ -364,20 +366,20 @@ EOF
	create_package "$fwpackagename" "$fwdir"
 fi

-cat <<EOF >> debian/control
-
-Package: $libc_headers_packagename
-Section: devel
-Provides: linux-kernel-headers
-Architecture: any
-Description: Linux support headers for userspace development
- This package provides userspaces headers from the Linux kernel.  These headers
- are used by the installed headers for GNU glibc and other system libraries.
-EOF
+#cat <<EOF >> debian/control
+#
+#Package: $libc_headers_packagename
+#Section: devel
+#Provides: linux-kernel-headers
+#Architecture: any
+#Description: Linux support headers for userspace development
+# This package provides userspaces headers from the Linux kernel.  These headers
+# are used by the installed headers for GNU glibc and other system libraries.
+#EOF

 if [ "$ARCH" != "um" ]; then
	create_package "$kernel_headers_packagename" "$kernel_headers_dir"
-	create_package "$libc_headers_packagename" "$libc_headers_dir"
+	#create_package "$libc_headers_packagename" "$libc_headers_dir"
 fi

 create_package "$packagename" "$tmpdir"
